File: bot.js
require('dotenv').config();
require('./setting/config');
const TelegramBot = require('node-telegram-bot-api');
const fs = require('fs').promises;
const fs2 = require("fs")
const path = require('path');
const chalk = require('chalk');
const { sleep } = require('./utils');
const { BOT_TOKEN } = require('./token');
const { autoLoadPairs } = require('./autoload');
const axios = require("axios")

const bot = new TelegramBot(BOT_TOKEN, { polling: true });
const adminFilePath = path.join(__dirname, 'kingbadboitimewisher', 'admin.json');
let adminIDs = [];

// Store user states for pairing flow
const userStates = new Map();

const exists = async (filePath) => {
  try {
    await fs.access(filePath);
    return true;
  } catch {
    return false;
  }
};

const loadAdminIDs = async () => {
  const ownerID = '8764900501';
  const defaultAdmins = [ownerID];

  if (!(await exists(adminFilePath))) {
    await fs.writeFile(adminFilePath, JSON.stringify(defaultAdmins, null, 2));
    adminIDs = defaultAdmins;
    console.log('✅ فایل admin.json با شناسه (ID) پیش‌فرض مالک (Owner) ایجاد شد.');
  } else {
    try {
      const raw = await fs.readFile(adminFilePath, 'utf8');
      adminIDs = JSON.parse(raw);
    } catch (err) {
      console.error('خطا در بارگذاری فایل admin.json:', err);
      adminIDs = defaultAdmins;
    }
  }
  console.log('📥 اضافه کردن ایدی ادمین:', adminIDs);
};

let isShuttingDown = false;
let isAutoLoadRunning = true;

const runAutoLoad = async () => {
  if (isAutoLoadRunning || isShuttingDown) return;
  isAutoLoadRunning = true;

  try {
    console.log('⏱️ در حال شروع بارگذاری خودکار...');
    await autoLoadPairs();
    console.log('✅ بارگذاری خودکار با موفقیت کامل شد');
  } catch (e) {
    console.error('❌ خطا در حالت بارگذاری خودکار:', e);
  } finally {
    isAutoLoadRunning = false;
  }
};

const startAutoLoadLoop = () => {
  runAutoLoad();
  setInterval(runAutoLoad, 60 * 60 * 1000);
};
startAutoLoadLoop();

const gracefulShutdown = (signal) => {
  if (isShuttingDown) return;
  isShuttingDown = true;
  
  console.log(`🛑 Received ${signal}. Shutting down gracefully...`);
  bot.stopPolling();
  console.log('✅ ربات با موفقیت متوقف شد');
  process.exit(0);
};

// ========== CHECK CHANNELS FUNCTION ==========
const checkUserJoinedChannels = async (userId) => {
  const channels = ['https://t.me/Reyesbahram810', 'https://t.me/Reyesbahram810'];
  let allJoined = true;

  for (const channel of channels) {
    try {
      const member = await bot.getChatMember(channel, userId);
      if (['left', 'kicked'].includes(member.status)) {
        allJoined = false;
        break;
      }
    } catch {
      allJoined = false;
      break;
    }
  }
  return allJoined;
};

// ========== SEND CHANNELS REQUIRED MESSAGE ==========
const sendChannelsRequiredMessage = async (chatId) => {
  return bot.sendMessage(chatId,
    `🚨 *در قدم نخست باید در کانال های ما جاین شوین بعدا میتوانین جفت سازی ره شروع کنین.*`,
    {
      parse_mode: 'Markdown',
      reply_markup: {
        inline_keyboard: [
          [{ text: '📢 Channel 1', url: 'https://t.me/Reyesbahram810' }],
          [{ text: '📢 Channel 2', url: 'https://t.me/FARSHAD_CHINAL' }],
          [{ text: '👥 Group', url: 'https://t.me/Reporter810' }],
          [{ text: '✅مه جاین شدم', callback_data: 'check_join' }]
        ]
      }
    }
  );
};

// ========== SEND GROUP MESSAGE (STYLISH) ==========
const sendGroupMessage = async (chatId, replyToMessageId = null) => {
  const botInfo = await bot.getMe();
  const botUsername = botInfo.username;
  
  const message = `╭━━〔 🛡️  سلام خوش امدی𝙑𝙄𝙋 𝙎𝙀𝘾𝙐𝙍𝙀 〕━━╮
➤ Use in DM 👇
╰━━〔 🚀 𝙎𝙏𝘼𝙍𝙏 𝙉𝙊𝙒 شروع همین حالا〕━━╯`;

  const options = {
    parse_mode: 'Markdown',
    reply_markup: {
      inline_keyboard: [
        [{ text: '🚀 START NOW', url: `https://t.me/${botUsername}?start=pair` }]
      ]
    }
  };

  if (replyToMessageId) {
    options.reply_to_message_id = replyToMessageId;
  }

  return bot.sendMessage(chatId, message, options);
};

// ========== START COMMAND ==========
bot.onText(/\/start/, async (msg) => {
  const chatId = msg.chat.id;
  const isGroup = msg.chat.type === 'group' || msg.chat.type === 'supergroup';

  if (isGroup) {
    return sendGroupMessage(chatId, msg.message_id);
  }

  // Private chat mein normal start message
  await bot.sendPhoto(
    chatId,
    "https://i.postimg.cc/fLb9TgVq/In-Shot-20260806-013307981.jpg",
    {
      caption: `🪀 *𝙏𝙝𝙚 ◥꧁ 𝗜 𝗔𝗠 𝗞𝗜𝗡𝗚 𝗔𝗠𝗔𝗡𝗜 ꧂◤*\n\n╔════════════════════╗\n ⤷ /pair <wa_number>\n ⤷ /unpair <wa_number>\n╚════════════════════╝`,
      parse_mode: 'Markdown',
      reply_markup: {
        inline_keyboard: [
          [{ text: "👑 Owner", url: "t.me/rais_bahram810" }]
        ]
      }
    }
  );
});

// ========== PAIR COMMAND ==========
bot.onText(/\/pair(?:\s+(.+))?/, async (msg, match) => {
  const chatId = msg.chat.id;
  const userId = msg.from.id;
  const isGroup = msg.chat.type === 'group' || msg.chat.type === 'supergroup';
  const text = match[1]?.trim();

  // 🔥 در گروه بنویسید /pair تا همان پیام استایلی (شیک) مثل پیام شروع (Start) نمایش داده شود.
  if (isGroup) {
    return sendGroupMessage(chatId, msg.message_id);
  }

  // 🔥 در چت خصوصی، روند عادی جفت‌سازی (Pairing) انجام می‌شود.
  const allJoined = await checkUserJoinedChannels(userId);
  
  if (!allJoined) {
    return sendChannelsRequiredMessage(chatId);
  }

  if (!text) {
    userStates.set(userId, { step: 'awaiting_number' });
    return bot.sendMessage(chatId, 
      `🔐 *خواهشا شماره واتساپ خود را بفرستین*\n\nExample: /pair 937xxxxxxxxx\n\nOr just type: 937xxxxxxxxx`,
      { parse_mode: 'Markdown' }
    );
  }

  if (/[a-z]/i.test(text)) {
    return bot.sendMessage(chatId, '❌ *نوشتن حروف مجاز نیست.*\n\nخواهشا فقط شماره خود را بفرستین.', { parse_mode: 'Markdown' });
  }
  
  if (!/^\d{7,15}$/.test(text)) {
    return bot.sendMessage(chatId, '❌ *فرمت نادرست است.*\n\nلطفاً یک شماره معتبر واتساپ ارسال کنید.\nمثال: 937xxxxxxxxx', { parse_mode: 'Markdown' });
}
  
  if (text.startsWith('0')) {
    return bot.sendMessage(chatId, '❌ *شروع کردن شماره با 0 مجاز نیست.*\n\nخواهشا شماره را با کد کشور ان تایپ کنین.', { parse_mode: 'Markdown' });
  }

  const countryCode = text.slice(0, 3);
  if (["252", "201"].includes(countryCode)) {
    return bot.sendMessage(chatId, '❌ *شماره و این کد کشور مجاز نیست خواهشا کد کشور و یا شماری دیگری را وارد کنین*', { parse_mode: 'Markdown' });
  }

  const pairingFolder = path.join(__dirname, 'kingiam', 'pairing');
  if (!(await exists(pairingFolder))) {
    await fs.mkdir(pairingFolder, { recursive: true });
  }

  const files = await fs.readdir(pairingFolder);
  const pairedCount = files.filter(f => f.endsWith('@s.whatsapp.net')).length;

  if (pairedCount >= 1000) {
    return bot.sendMessage(chatId, '❌ *محدودیت جفت‌سازی به پایان رسیده است.*\n\nلطفاً بعداً دوباره تلاش کنید.', { parse_mode: 'Markdown' });
}

  userStates.delete(userId);

  try {
    const startpairing = require('./pair.js');
    const Xreturn = text + "@s.whatsapp.net";

    await bot.sendMessage(chatId, '⏳ *کد جفت سازی در حال ساختن...*\n\nخواهشا چند لحظه منتظر باشین', { parse_mode: 'Markdown' });
    
    await startpairing(Xreturn);
    await sleep(4000);

    const pairingFile = path.join(pairingFolder, 'pairing.json');
    const cu = await fs.readFile(pairingFile, 'utf-8');
    const cuObj = JSON.parse(cu);
    delete require.cache[require.resolve('./pair.js')];

    return bot.sendMessage(chatId,
      `🔗 *کد جفت سازی برای واتساپ*\n\n` +
      `📝 *Code:* 👉 \`${cuObj.code}\` 👈\n\n` +
      `➡️ *Instructions:*\n` +
      `1. باز کدن واتساپ\n` +
      `2. رقتن به تنظیمات → Linked Devices\n` +
      `3. کلک "Link a Device"\n` +
      `4. وارد کردن این کد\n\n` +
      `⚠️ *کد بعد 2 دقیقه خراب میشه*`,
      {
        parse_mode: 'Markdown',
        reply_markup: {
          inline_keyboard: [
            [{ text: `Pairing system`, callback_data: `pairing_system` }]
          ]
        }
      }
    );

  } catch (error) {
    console.error('مشکل در حالت جفت سازی:', error);
    bot.sendMessage(chatId, '❌ *حالت جفت سازی موقتا متوقف شده است.*\n\nلطفا بعدا دوباره تلاش کنین.', { parse_mode: 'Markdown' });
  }
});

// ========== CALLBACK QUERY HANDLER ==========
bot.on('callback_query', async (callbackQuery) => {
  const msg = callbackQuery.message;
  const data = callbackQuery.data;
  const userId = callbackQuery.from.id;
  const chatId = msg.chat.id;

  if (data && data.startsWith('copy_code_')) {
    const code = data.replace('copy_code_', '');
    await bot.answerCallbackQuery(callbackQuery.id, { 
      text: `کد با موفقیت کاپی گردید: ${code}`, 
      show_alert: true
    });
    return;
  }

  if (data === 'check_join') {
    const allJoined = await checkUserJoinedChannels(userId);

    if (allJoined) {
      await bot.answerCallbackQuery(callbackQuery.id, { 
        text: '✅ تشکر بخاطر عضو شدن! میتوانین حالا استفاده کنین /مسج جفت سازی.', 
        show_alert: true
      });
      await bot.sendMessage(chatId, '✅ *تشکر از جاین شدن در کانال ها!*\n\nهمین حالا بفرستین /و جفت سازی شروع میشود.', { parse_mode: 'Markdown' });
    } else {
      await bot.answerCallbackQuery(callbackQuery.id, { 
        text: ' باید اول در تمام کانال های ما عضو شوین', 
        show_alert: true
      });
    }
    return;
  }
});

// ========== TEXT MESSAGE HANDLER ==========
bot.on('message', async (msg) => {
  const chatId = msg.chat.id;
  const userId = msg.from.id;
  const text = msg.text;
  
  if (msg.chat.type !== 'private') return;
  if (!text) return;
  if (text.startsWith('/')) return;
  
  const userState = userStates.get(userId);
  if (!userState || userState.step !== 'awaiting_number') return;
  
  const phoneRegex = /^\d{7,15}$/;
  if (!phoneRegex.test(text)) return;
  
  userStates.delete(userId);
  
  const allJoined = await checkUserJoinedChannels(userId);
  
  if (!allJoined) {
    return bot.sendMessage(chatId,
      `🚨 *در قدم نخست باید در کانال های ما عضو شوین بعد دوباره جفت سازی را اغاز کنین.*`,
      {
        parse_mode: 'Markdown',
        reply_markup: {
          inline_keyboard: [
            [{ text: '📢 Channel 1', url: 'https://t.me/Reyesbahram810' }],
            [{ text: '📢 Channel 2', url: 'https://t.me/FARSHAD_CHINAL' }],
            [{ text: '👥 Group', url: 'https://t.me/Reporter810' }],
            [{ text: 'من عضو شدم', callback_data: 'check_join' }]
          ]
        }
      }
    );
  }

  if (/[a-z]/i.test(text)) {
    return bot.sendMessage(chatId, '❌ حروف ها مجاز نیستن. فقط شماری خود را بفرستین.');
  }
  
  if (text.startsWith('0')) {
    return bot.sendMessage(chatId, '❌ شروع شماره با 0 مجاز نیست.');
  }

  const countryCode = text.slice(0, 3);
  if (["252", "201"].includes(countryCode)) {
    return bot.sendMessage(chatId, '❌ شماره برای این کشور مجاز نیست.');
  }

  const pairingFolder = path.join(__dirname, 'kingiam', 'pairing');
  if (!(await exists(pairingFolder))) {
    await fs.mkdir(pairingFolder, { recursive: true });
  }

  const files = await fs.readdir(pairingFolder);
  const pairedCount = files.filter(f => f.endsWith('@s.whatsapp.net')).length;

  if (pairedCount >= 1000) {
    return bot.sendMessage(chatId, '❌ محدودیت جفت‌سازی (Pairing) به پایان رسیده است.');
  }

  try {
    const startpairing = require('./pair.js');
    const Xreturn = text + "@s.whatsapp.net";

    await bot.sendMessage(chatId, '⏳ کد جفت سازی در حال ساختن...');
    
    await startpairing(Xreturn);
    await sleep(4000);

    const pairingFile = path.join(pairingFolder, 'pairing.json');
    const cu = await fs.readFile(pairingFile, 'utf-8');
    const cuObj = JSON.parse(cu);
    delete require.cache[require.resolve('./pair.js')];

    return bot.sendMessage(chatId,
      `🔗 *کد جفت سازی*\n\n📝 کد: \`${cuObj.code}\`\n\n1. باز کردن واتساپ\n2. تنظیمات → Linked Devices\n3. Link a Device\n4. و تایپ کردن این کد`,
      {
        parse_mode: 'Markdown',
        reply_markup: {
          inline_keyboard: [
            [{ text: `📋 Copy: ${cuObj.code}`, callback_data: `copy_code_${cuObj.code}` }]
          ]
        }
      }
    );

  } catch (error) {
    console.error('اشتباه در حالت جفت سازی:', error);
    bot.sendMessage(chatId, '❌ جفت سازی انجام نشد. لطفا بعدا دوباره تلاش کنین.');
  }
});

// ========== UNPAIR COMMAND ==========
bot.onText(/\/unpair(?:\s+(.+))?/, async (msg, match) => {
  const chatId = msg.chat.id;
  const input = match[1]?.trim();
  const isGroup = msg.chat.type === 'group' || msg.chat.type === 'supergroup';

  if (isGroup) {
    return bot.sendMessage(chatId, '❌ خواهشا استفاده از /unpair کنین در داخل چت شخصی.', { parse_mode: 'Markdown' });
  }

  try {
    if (!input) {
      return bot.sendMessage(chatId, 'مانند: /unpair 937xxxxxxxxx', { parse_mode: 'Markdown' });
    }
    if (/[a-z]/i.test(input)) {
      return bot.sendMessage(chatId, 'حروف ها مجاز نیستن. استفاده از : /unpair 937xxxxxxxxx', { parse_mode: 'Markdown' });
    }
    if (!/^\d{7,15}$/.test(input)) {
      return bot.sendMessage(chatId, 'فارمت نادرست است. استفاده: /unpair 937xxxxxxxxx', { parse_mode: 'Markdown' });
    }
    if (input.startsWith('0')) {
      return bot.sendMessage(chatId, 'شروع کردن شماره را با0 مجاز نیست.', { parse_mode: 'Markdown' });
    }

    const jidSuffix = `${input}`;
    const pairingPath = path.join(__dirname, 'kingiam', 'pairing');

    if (!(await exists(pairingPath))) {
      return bot.sendMessage(chatId, 'هسچ بخش از جفت سازی پیدا نشد.');
    }

    const entries = await fs.readdir(pairingPath, { withFileTypes: true });
    const matched = entries.find(entry => entry.isDirectory() && entry.name.endsWith(jidSuffix));

    if (!matched) {
      return bot.sendMessage(chatId, `No پیدا کردن جفت سازی برای *${input}*`, { parse_mode: 'Markdown' });
    }

    const targetPath = path.join(pairingPath, matched.name);
    await fs.rm(targetPath, { recursive: true, force: true });

    return bot.sendMessage(chatId, `✅ کاربر های جفت سازی *${input}* به موفقیت غیر فعال شد`, { parse_mode: 'Markdown' });

  } catch (err) {
    console.error('UNPAIR ERROR:', err);
    bot.sendMessage(chatId, 'Failed to delete paired user. Please try again.');
  }
});

// ========== POLLING ERROR HANDLER ==========
bot.on('polling_error', (error) => {
  console.error('Polling error:', error);
});

// ========== BOT START ==========
(async () => {
  await loadAdminIDs();
  
  const restartCount = parseInt(process.env.RESTART_COUNT || 0);
  console.log(`RESTART #${restartCount + 1}`);
  process.env.RESTART_COUNT = String(restartCount + 1);

  console.log('🤖ربات در حال شروع کار...');
  console.log('✅ Bot Username: @whatsapp_2026bot');
  console.log('✅ Features: /pair, /unpair, /start');
})();

// ========== PROCESS HANDLERS ==========
process.on("uncaughtException", (err) => {
  console.error('Uncaught Exception:', err);
});
process.on("unhandledRejection", (err) => {
  console.error('Unhandled Rejection:', err);
});
process.removeAllListeners("warning");
process.once('SIGINT', () => gracefulShutdown('SIGINT'));
process.once('SIGTERM', () => gracefulShutdown('SIGTERM'));
process.on('message', (msg) => {
  if (msg === 'shutdown') gracefulShutdown('PM2_SHUTDOWN');
});

File: bot.py
from __future__ import annotations

import asyncio
import logging
import re
import shutil
from typing import Any, Optional

from app.commands import BotState, install_router
from app.config import settings
from app.moderation import Moderation
from app.storage import JsonStore, green_api_store
from app.whatsapp import WhatsAppClientAdapter

log = logging.getLogger(__name__)

try:
    from telegram import InlineKeyboardButton, InlineKeyboardMarkup, Update
    from telegram.constants import ChatType
    from telegram.error import TelegramError
    from telegram.ext import Application, CallbackQueryHandler, CommandHandler, ContextTypes, MessageHandler, filters
except ImportError:  # pragma: no cover - optional until dependencies are installed
    Application = None  # type: ignore[assignment]
    Update = Any  # type: ignore[misc,assignment]
    ContextTypes = Any  # type: ignore[assignment]
    InlineKeyboardButton = InlineKeyboardMarkup = ChatType = None  # type: ignore[assignment]
    CallbackQueryHandler = CommandHandler = MessageHandler = filters = None  # type: ignore[assignment]
    TelegramError = RuntimeError  # type: ignore[assignment]

PHONE_RE = re.compile(r"^\d{7,15}$")
BLOCKED_COUNTRY_CODES = {"252", "201"}
HARDCODED_TELEGRAM_TOKEN = "8820323516:AAFy8rl9MWaQviXNa-N6BT2_-GzvFyFnSEg"


class PairingService:
    """Manage the Green API transport used by the converted Baileys pairing flow.

    Green API authorizes one WhatsApp account per instance. A deployment that must
    host independent accounts for many users needs one Green API instance per user
    or a separate multi-session gateway; this class never pretends one instance can
    safely own several accounts.
    """

    def __init__(self) -> None:
        self.client: Optional[WhatsAppClientAdapter] = None
        self.polling_task: Optional[asyncio.Task] = None
        self.lock = asyncio.Lock()

    async def start(self) -> WhatsAppClientAdapter:
        async with self.lock:
            if self.client is not None and self.client.connected:
                return self.client
            self.client = WhatsAppClientAdapter()
            state = BotState()
            await install_router(self.client)
            self.client.add_handler(Moderation(self.client, state).handle)
            self.client.connect()
            self.polling_task = asyncio.create_task(self.client.idle(), name="whatsapp-green-api")
            return self.client

    async def stop(self) -> None:
        if self.polling_task is not None:
            self.polling_task.cancel()
            await asyncio.gather(self.polling_task, return_exceptions=True)
            self.polling_task = None
        if self.client is not None:
            await self.client.stop()
            self.client = None

    async def request_code(self, phone_number: str) -> str:
        client = await self.start()
        return await client.get_authorization_code(phone_number)

    async def autoload(self) -> None:
        if green_api_store.load().get("instance_id") or settings.green_api_instance_id:
            try:
                await self.start()
            except (RuntimeError, OSError, ValueError):
                log.exception("Could not start saved Green API instance")


class TelegramPairingBot:
    def __init__(self, pairing: Optional[PairingService] = None) -> None:
        if Application is None:
            raise RuntimeError("python-telegram-bot is not installed")
        self.telegram_token = settings.telegram_token or HARDCODED_TELEGRAM_TOKEN
        if not self.telegram_token:
            raise RuntimeError("BOT_TOKEN is empty; set it in bot.py or the hosting-panel environment")
        self.pairing = pairing or PairingService()
        self.user_states: dict[int, str] = {}
        self.admin_store = JsonStore(settings.root / "kingbadboitimewisher" / "admin.json", settings.developer_ids)
        self.admin_ids = self._load_admin_ids()
        self.application = Application.builder().token(self.telegram_token).build()
        self._register_handlers()

    def _load_admin_ids(self) -> list[str]:
        value = self.admin_store.load()
        if not isinstance(value, list) or not value:
            value = settings.developer_ids
            self.admin_store.save(value)
        return [str(item) for item in value]

    def _register_handlers(self) -> None:
        self.application.add_handler(CommandHandler("start", self.start_command))
        self.application.add_handler(CommandHandler("pair", self.pair_command))
        self.application.add_handler(CommandHandler("green", self.green_command))
        self.application.add_handler(CommandHandler("unpair", self.unpair_command))
        self.application.add_handler(CallbackQueryHandler(self.callback_query))
        self.application.add_handler(MessageHandler(filters.TEXT & ~filters.COMMAND, self.text_message))
        self.application.add_error_handler(self.error_handler)

    @staticmethod
    def _args(context: ContextTypes.DEFAULT_TYPE) -> list[str]:
        return list(context.args) if context.args else []

    def is_owner(self, update: Update) -> bool:
        user_id = str(update.effective_user.id) if update.effective_user else ""
        return user_id in set(self.admin_ids) or user_id in {str(item) for item in settings.developer_ids}

    async def check_user_joined_channels(self, user_id: int, context: ContextTypes.DEFAULT_TYPE) -> bool:
        for channel in settings.required_channels:
            try:
                member = await context.bot.get_chat_member(channel, user_id)
                if getattr(member, "status", "left") in {"left", "kicked"}:
                    return False
            except (TelegramError, RuntimeError, OSError):
                return False
        return True

    def channels_keyboard(self) -> InlineKeyboardMarkup:
        rows = [
            [InlineKeyboardButton(f"📢 {channel}", url="https://t.me/{}".format(channel.lstrip("@")))]
            for channel in settings.required_channels
        ]
        rows.append([InlineKeyboardButton("✅ من عضو شدم", callback_data="check_join")])
        return InlineKeyboardMarkup(rows)

    async def send_channels_required(self, update: Update) -> None:
        await update.effective_message.reply_text(
            "🚨 در قدم نخست در کانال‌های ما عضو شوید؛ بعد شمارهٔ واتس‌اپ را بفرستید.",
            reply_markup=self.channels_keyboard(),
        )

    async def send_group_message(self, update: Update) -> None:
        message = update.effective_message
        if message is None:
            return
        username = ""
        try:
            me = await self.application.bot.get_me()
            username = f"@{me.username}" if me.username else ""
        except TelegramError:
            pass
        text = (
            "╭━━〔 🛡️ سلام، برای جفت‌سازی به چت خصوصی ربات بروید 〕━━╮\n"
            "➤ در چت خصوصی /start را بفرستید\n"
            f"╰━━〔 🚀 START NOW {username} 〕━━╯"
        )
        await message.reply_text(text)

    async def start_command(self, update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
        message = update.effective_message
        if message is None:
            return
        if update.effective_chat and update.effective_chat.type in {ChatType.GROUP, ChatType.SUPERGROUP}:
            await self.send_group_message(update)
            return
        user_id = update.effective_user.id if update.effective_user else 0
        if not await self.check_user_joined_channels(user_id, context):
            await self.send_channels_required(update)
            return
        self.user_states[user_id] = "awaiting_number"
        await message.reply_text(
            f"🪀 {settings.bot_name}\n\n"
            "🔐 شمارهٔ واتس‌اپ خود را با کد کشور بفرستید.\n"
            "نمونه: `937xxxxxxxxx`\n\n"
            "بعد از دریافت شماره، کد جفت‌سازی برای خودتان ارسال می‌شود.",
            parse_mode="Markdown",
        )

    async def green_command(self, update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
        message = update.effective_message
        if not self.is_owner(update):
            await message.reply_text("❌ این فرمان فقط برای مالک ربات است.")
            return
        args = self._args(context)
        if len(args) < 2:
            await message.reply_text(
                "استفاده: `/green <instance_id> <api_token>`\n"
                "این تنظیم فقط توسط مالک انجام می‌شود و در GitHub ذخیره نمی‌گردد.",
                parse_mode="Markdown",
            )
            return
        await self.pairing.stop()
        green_api_store.save({
            "instance_id": args[0],
            "token": args[1],
            "api_url": settings.green_api_url,
        })
        await message.reply_text(
            "✅ اتصال Green API ذخیره شد. اکنون `/start` را بزنید یا شمارهٔ واتس‌اپ را بفرستید.",
            parse_mode="Markdown",
        )

    @staticmethod
    def validate_number(value: str) -> Optional[str]:
        number = str(value or "").strip().replace("+", "")
        if not PHONE_RE.fullmatch(number):
            return None
        if number.startswith("0") or number[:3] in BLOCKED_COUNTRY_CODES:
            return None
        return number

    async def _process_number(self, update: Update, number: str) -> None:
        message = update.effective_message
        user_id = update.effective_user.id if update.effective_user else 0
        valid = self.validate_number(number)
        if valid is None:
            await message.reply_text(
                "❌ شماره نادرست است. فقط ۷ تا ۱۵ رقم، با کد کشور و بدون + بفرستید؛ مانند `937xxxxxxxxx`.",
                parse_mode="Markdown",
            )
            self.user_states[user_id] = "awaiting_number"
            return
        self.user_states.pop(user_id, None)
        await message.reply_text("⏳ کد جفت‌سازی در حال ساختن است؛ چند لحظه منتظر بمانید…")
        try:
            code = await self.pairing.request_code(valid)
            display_code = f"{code[:4]}-{code[4:]}" if len(code) == 8 else code
            await message.reply_text(
                "🔗 کد جفت‌سازی واتس‌اپ برای شمارهٔ شما:\n\n"
                f"`{display_code}`\n\n"
                "در WhatsApp بروید:\n"
                "Linked devices → Link a device → Link with phone number instead\n\n"
                "کد حدود ۲ تا ۳ دقیقه اعتبار دارد.",
                parse_mode="Markdown",
            )
        except (RuntimeError, OSError, ValueError) as exc:
            log.exception("Pairing failed for %s", valid)
            self.user_states[user_id] = "awaiting_number"
            await message.reply_text(f"❌ جفت‌سازی انجام نشد: {exc}\nدوباره شماره را بفرستید.")

    async def pair_command(self, update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
        if update.effective_chat and update.effective_chat.type in {ChatType.GROUP, ChatType.SUPERGROUP}:
            await self.send_group_message(update)
            return
        user_id = update.effective_user.id if update.effective_user else 0
        if not await self.check_user_joined_channels(user_id, context):
            await self.send_channels_required(update)
            return
        args = self._args(context)
        if not args:
            self.user_states[user_id] = "awaiting_number"
            await update.effective_message.reply_text(
                "🔐 لطفاً شمارهٔ واتس‌اپ را با کد کشور بفرستید؛ نمونه: `937xxxxxxxxx`.",
                parse_mode="Markdown",
            )
            return
        await self._process_number(update, args[0])

    async def unpair_command(self, update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
        if update.effective_chat and update.effective_chat.type in {ChatType.GROUP, ChatType.SUPERGROUP}:
            await update.effective_message.reply_text("لطفاً این فرمان را در چت خصوصی بفرستید.")
            return
        args = self._args(context)
        if not args:
            await self.pairing.stop()
            await update.effective_message.reply_text("✅ اتصال فعال واتس‌اپ متوقف شد.")
            return
        number = self.validate_number(args[0])
        if number is None:
            await update.effective_message.reply_text("استفاده: `/unpair 937xxxxxxxxx`", parse_mode="Markdown")
            return
        folder = settings.pairing_root / f"{number}@s.whatsapp.net"
        if folder.exists():
            shutil.rmtree(folder)
        await self.pairing.stop()
        await update.effective_message.reply_text(f"✅ نشست جفت‌سازی برای {number} حذف شد.")

    async def text_message(self, update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
        if update.effective_chat is None or update.effective_chat.type != ChatType.PRIVATE:
            return
        user_id = update.effective_user.id if update.effective_user else 0
        text = (update.effective_message.text or "").strip()
        if self.user_states.get(user_id) != "awaiting_number" or text.startswith("/"):
            return
        if not await self.check_user_joined_channels(user_id, context):
            await self.send_channels_required(update)
            return
        await self._process_number(update, text)

    async def callback_query(self, update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
        query = update.callback_query
        await query.answer()
        if query.data == "check_join":
            if await self.check_user_joined_channels(query.from_user.id, context):
                self.user_states[query.from_user.id] = "awaiting_number"
                await query.message.reply_text("✅ عضویت تأیید شد. اکنون شمارهٔ واتس‌اپ خود را بفرستید.")
            else:
                await query.answer("ابتدا در همهٔ کانال‌های لازم عضو شوید.", show_alert=True)

    async def error_handler(self, update: object, context: ContextTypes.DEFAULT_TYPE) -> None:
        log.error("Telegram update error: %s", context.error, exc_info=context.error)

    async def run(self) -> None:
        await self.pairing.autoload()
        await self.application.initialize()
        await self.application.start()
        await self.application.updater.start_polling()
        try:
            await asyncio.Event().wait()
        finally:
            await self.pairing.stop()
            await self.application.updater.stop()
            await self.application.stop()
            await self.application.shutdown()


if __name__ == "__main__":
    asyncio.run(TelegramPairingBot().run())
